### PR TITLE
Bats

### PR DESCRIPTION
**Pull request by silverdaz**
Setting the stage for BATS

### Describe the pull request:

- [x] Bug fix
- [x] Functional change
- [ ] New feature
- [x] Code cleanup
- [x] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
Cleaning up the old tests and setting up BATS

### Related issues:
#17

### Mentions:
@omllobet

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -16,15 +16,21 @@ deps = flake8
 commands = flake8 .
 
 [testenv:unit_tests]
+deps = coverage
+       -rrequirements.txt
+       -rtests/unit/requirements.txt
+commands = coverage run -m pytest -x tests/unit
+	   coverage report
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tox.ini]: Done.
Omission stated in the .coveragerc file

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  Expected outcome: Authentication "fails", and the ingestion does not start
+
+* Ingest a file with a user in CentralEGA, using the wrong password<br/>
+  Expected outcome: Authentication "fails", and the ingestion does not start
+
+* Ingest a file with a user in CentralEGA, using the wrong sshkey<br/>
+  Expected outcome: Fallback to password (previous scenario)
+
+* Ingest a file for a given LocalEGA using the key of another one<br/>
+  Expected outcome: Message in the CentralEGA error queue, with the relevant content.
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/README.md]: You are confusing 2 cases.
* The user encrypts a file using a pgp key, ie. for some LocalEGA instance, and uploads it to another LocalEGA instance. The latter should "refuse" the file, and an error should be redirected to the user.
* The user tries to upload a file to some LocalEGA instance using inproper credentials.

File: tests/integration/inbox.bats [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    # Defining the TMP dir
+    TESTFILES=${BATS_TEST_FILENAME}.d
+    mkdir -p $TESTFILES
+
+    # Test user
+    TESTUSER=dummy
+
+    # Utilities to scan the Message Queues
+    MQ_CONSUME="python ${MAIN_REPO}/extras/rabbitmq/consume.py --connection ${CEGA_CONNECTION}"
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/integration/inbox.bats]: Good point. I started migrating other scripts there, so I could do the same with the rabbitmq ones

File: tests/robustness/db.bats [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    MQ_PUBLISH="python ${MAIN_REPO}/extras/rabbitmq/publish.py --connection ${CEGA_CONNECTION}"
+
+    # Find inbox port mapping. Usually 2222:9000
+    legarun docker port inbox 9000
+    [ "$status" -eq 0 ]
+    INBOX_PORT=${output##*:}
+    LEGA_SFTP="sftp -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -P $INBOX_PORT"
+}
+
+function teardown() {
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/robustness/db.bats]: True, but I wanted the flexibility to redefine the function in the current bats test if necessary.
So far, you're right, I noticed it myself, it's not very used. But I don't know about future test cases. So I left it.

I'm up for refactoring a bit to make it look better. The functionality is there though.

File: tests/README.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
+# LocalEGA testsuite
+
+Unit Tests are run with pytest, coverage and tox.
+The other tests use BATS.
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/README.md]: Done

File: tests/integration/inbox.bats (hunk 3 above)
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/integration/inbox.bats]: Utilities migrated to `tests/_common/mq`

File: deploy/Makefile [hunk 6]
```diff
@@ -35,10 +25,37 @@ ps:
...
+
+# For the moment, we simply do sleep 60, because we need
+# - the inbox notification server up
+# - the rabbitmq shovel to CentralEGA (the federated queue can be late, it doesn't hurt)
+# - the key server (the java spring needs 97 seconds, 10 are enough for the other)
+
+
+# Decrypt the encoded data from the repository
+# That is, some users private ssh keys and the database of fake CentralEGA Users
+prepare: ../tests/_common/dummy.sec ../tests/_common/john.sec ../tests/_common/jane.sec ../tests/_common/users.json
```
Comment (blankdots) [inline deploy/Makefile]: maybe a different solution to prepare with `openssl` as I get:
```
╰─$ make prepare                                                                                                                                                                                               2 ↵
openssl enc -aes-256-cbc -d -in ../tests/_common/john.sec.enc -out ../tests/_common/john.sec -pass pass:1PbDCkTX5y8BnOjw
*** WARNING : deprecated key derivation used.
Using -iter or -pbkdf2 would be better.
bad decrypt
139748720580416:error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:crypto/evp/evp_enc.c:537:
make: *** [../tests/_common/john.sec] Error 1

```

File: tests/integration/ingestion.bats [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    retry_until 0 100 1 ${MQ_GET} v1.files.inbox "${TESTUSER}" "/${TESTFILE}.c4ga"
+    [ "$status" -eq 0 ]
+    CORRELATION_ID=$output
+
+    # Publish the file to simulate a CentralEGA trigger
+    MESSAGE="{ \"user\": \"${TESTUSER}\", \"filepath\": \"/${TESTFILE}.c4ga\"}"
+    legarun ${MQ_PUBLISH} --correlation_id ${CORRELATION_ID} files "$MESSAGE"
+    [ "$status" -eq 0 ]
+
+    # Check that a message with the above correlation id arrived in the expected queue
```
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/integration/ingestion.bats]: nope, not here.
If you want to make such a test, it's ok, but it'll have to be in another place.
I can think of putting it in the robustness part.
I'll explain more tomorrow.

File: deploy/Makefile (hunk 6 above)
Comment (silverdaz) [reply deploy/Makefile]: Yes. Oscar had the same problem.
As this is a convenience target, please suggest a way to provide the files to the rest of the team. I can create a zip file and send it to Slack, and remove those lines. 
What do you say?

Comment (blankdots) [reply deploy/Makefile]: I say tests need to be replicated by anyone cloning the repo, and I admit I have no good solution at this point for encrypting those files, besides using maybe encrypting with crypt4gh and having the key attached to travis in base64 maybe ... what say you ?

File: tests/integration/ingestion.bats (hunk 7 above)
[the remark this answers is not in the document]
Comment (blankdots) [reply tests/integration/ingestion.bats]: The idea behind such a db tests is that in unit tests we tested the db functions, but we have not checked that the data is registered as we expect in the actual PostgresDB, thus we need to check it either in ingestion (that is why I suggested here) or somewhere else

File: tests/robustness/db.bats (hunk 4 above)
[the remark this answers is not in the document]
Comment (blankdots) [reply tests/robustness/db.bats]: We can refactor more later .. if need be

**Comment by blankdots** [comment]
I approve of this PR from a code base perspective.

This indeed provides faster run times of the tests, but we shall see how these times keep their "fast" status over time. Another aspect is that is minimizes the number of dependencies needed to run the tests.

Looking at things objectively, maybe investing the time to refresh the previous tests would have achieved the same result, instead of removing them... my understanding is that BATS supports Java and similar functionality could have been achieved, and we only would have needed to add the new tests

I strongly recommend a better preflight-check mechanism in a later PR with the https://github.com/jwilder/dockerize in a container, although it is not complicated to do it, having it part of this PR would have things complicated to review.

I hope these test cases can be extended to run on all deployment environments (kube, swarm etc.), and by the looks of it it seems that is feasible.

File: .travis.yml [hunk 8]
```diff
@@ -1,80 +1,79 @@
...
-        - pip install git+https://github.com/NBISweden/LocalEGA-cryptor.git
-        - pip install pika
-        - make -C test
-        - sleep 10
-        - make -C test check
+        - cd ../tests
+        - bats security
+    - stage: images
+      name: "Images Build"
+      script: echo 'Next PR'
```
Comment (blankdots) [inline .travis.yml]: I assume Image build will be part of a different PR ?

File: deploy/Makefile (hunk 6 above)
Comment (silverdaz) [reply deploy/Makefile]: hehe. Well. Let's not involve crypt4gh here.
I think we can figure something with openssl.
I'll pass some flag for better key derivation function, etc...
Can we do it in another PR? (since it currently works on Travis, and I can provide the files to the team)

File: .travis.yml (hunk 8 above)
Comment (silverdaz) [reply .travis.yml]: Correct. 
I have a description in the presentation for tomorrow, and I'm involving the team to take a decision.

File: tests/integration/ingestion.bats (hunk 7 above)
[the remark this answers is not in the document]
Comment (silverdaz) [reply tests/integration/ingestion.bats]: In terms of the logic involved in testing, you don't check that the db receives the right value. You check that whichever component that receives/needs the value can do its function.
The value itself is irrelevant.

Assume that you change the component to request another value, and you also change the database to store the new value, then in that case, the functionality is there and all fine, but the test fails. The issue is in the test.

The test should tell you whether an _implementation_ matches a _specification_.
When the implementation changes, but not the specification, there is no reason to change the test. If this is needed, the test is flawed.

That said, there are some "checking" that can be done, to see if the database does store some appropriate selected values, and in that case, I'd say it's a unit test for the database, not an integration test. Integration tests are concerned with specifications, not implementations.

[the remark this answers is not in the document]
Comment (blankdots) [reply tests/integration/ingestion.bats]: that sounds more like Functional tests rather than Integration tests, I view integration tests as focused also on that components integrate with one another meaning looking that the internals are ok ... but then again these views are for a Slack discussion.

File: deploy/Makefile (hunk 6 above)
Comment (blankdots) [reply deploy/Makefile]: > Can we do it in another PR?

We can, but I would personally prefer if things are nice and work once we merge.

**Comment by silverdaz** [comment]
> Rebase and remove the merge commits.

Merge commits removed.

**Comment by silverdaz** [comment]
I understand that.
In our case, we need some private keys for the ssh connections (here sftp).
There was no solution to include the private key in Travis, and it doesn't work to add it in some env variable (issues with escaping/quoting).
I reseached how to solve it, and it is best practice to do so, when you have to use private keys on Travis: You encrypt them and pass the password to travis in some env variable.
An arbitrary developer does not have the decryption password.

Do you have a another solution? A better solution?

**Comment by jbygdell** [comment]
1# We should test connecting to CEGA using both password and ssh_key as it is supported.

2# Ssh keys can be placed in a private repo and pulled from there by travis, 
that way we don't have to have live keys in  a public repo.
The keys will then be available for project members but not the general public.

**Comment by silverdaz** [comment]
Sounds good.
How does Travis pull from the private repo, a deploy-key thingy?

**Comment by jbygdell** [comment]
Either that or supply user:pass when cloning via https

File: deploy/Makefile [hunk 9]
```diff
@@ -1,32 +1,22 @@
...
-.PHONY: help bootstrap private network up down clean ps clean-volumes clean-all
+.PHONY: help bootstrap private network up down clean ps clean-volumes clean-all prepare
 
 help:
 	@echo "Usage: make <target>\n"
 	@echo "where <target> is: 'bootstrap', 'up' 'ps', 'down', 'network' or 'clean'\n"
 
 # If DEPLOY_DEV is yes, we use dummy passwords
-bootstrap: ENVS=--env DEPLOY_DEV=yes
-private/cega.yml private/lega.yml private bootstrap:
```
Comment (dtitov) [inline deploy/Makefile]: I personally don't like the idea of removing `cega.yml`. It's confusing that in case of spinning up local fake-CEGA we have definitions for it in `lega.yml`.

File: deploy/Makefile (hunk 6 above)
Comment (blankdots) [reply deploy/Makefile]: In theory that `make prepare` should work for everybody, @silverdaz please let us have a better solution in a following PR.

**Comment by blankdots** [comment]
Seems that all comments are addressed now or to be solved in following PR(s), merging ...
